MuxExoPlayer/src/main/java/com/mux/stats/sdk/muxstats/MuxBaseExoPlayer.java

```diff
@@ -779,251 +787,214 @@ public void outputLog(String tag, String msg) {
 
   class BandwidthMetric {
 
-    public BandwidthMetricData onLoadError(DataSpec dataSpec, int dataType, IOException e) {
-      BandwidthMetricData loadData = new BandwidthMetricData();
-      loadData.setRequestError(e.toString());
-      if (dataSpec != null && dataSpec.uri != null) {
-        loadData.setRequestUrl(dataSpec.uri.toString());
-        loadData.setRequestHostName(dataSpec.uri.getHost());
-      }
-      switch (dataType) {
-        case C.DATA_TYPE_MANIFEST:
-          loadData.setRequestType("manifest");
-          break;
-        case C.DATA_TYPE_MEDIA:
-          loadData.setRequestType("media");
-          break;
-        default:
-          return null;
-      }
-      loadData.setRequestErrorCode(null);
-      loadData.setRequestErrorText(e.getMessage());
-      return loadData;
-    }
+    TrackGroupArray availableTracks;
+    HashMap<String, BandwidthMetricData> loadedSegments = new HashMap<>();
 
-    public BandwidthMetricData onLoadCanceled(DataSpec dataSpec) {
-      BandwidthMetricData loadData = new BandwidthMetricData();
-      loadData.setRequestCancel("genericLoadCanceled");
-      if (dataSpec != null && dataSpec.uri != null) {
-        loadData.setRequestUrl(dataSpec.uri.toString());
-        loadData.setRequestHostName(dataSpec.uri.getHost());
+    public BandwidthMetricData onLoadError(String segmentUrl,
+        IOException e) {
+      BandwidthMetricData segmentData = loadedSegments.get(segmentUrl);
+      if (segmentData == null) {
+        segmentData = new BandwidthMetricData();
+        // TODO We should see how to put minimal stats here !!!
       }
-      loadData.setRequestType("media");
-      return loadData;
+      segmentData.setRequestError(e.toString());
+      // TODO see what error codes are
+      segmentData.setRequestErrorCode(-1);
+      segmentData.setRequestErrorText(e.getMessage());
+      segmentData.setRequestResponseEnd(System.currentTimeMillis());
+      return segmentData;
+    }
+
+    public BandwidthMetricData onLoadCanceled(String segmentUrl) {
+      BandwidthMetricData segmentData = loadedSegments.get(segmentUrl);
+      if (segmentData == null) {
+        segmentData = new BandwidthMetricData();
+        // TODO We should see how to put minimal stats here !!!
```

Comment:
Is action needed here?

Reply:
Nothing to do there, i9f that case happen we did not collect the data correctly. Maybe some dummy datas can be put in this case.

Reply:
Do you have any suggestions for minimal stats we could include?

Reply:
None that make sense, what can be calculated will be sent, what we missed in this case is segment start stats, we can only make it up but it will not give a valid data.